src/platforms/linux/linuxpingsender.cpp

```diff
@@ -19,12 +21,52 @@ namespace {
 Logger logger({LOG_LINUX, LOG_NETWORKING}, "LinuxPingSender");
 }
 
+int LinuxPingSender::createSocket() {
+  // Try creating an ICMP socket. This would be the ideal choice, but it can
+  // fail depending on the kernel config (see: sys.net.ipv4.ping_group_range)
+  m_socket = socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP);
+  if (m_socket >= 0) {
+    m_ident = 0;
+    return m_socket;
+  }
+  if ((errno != EPERM) && (errno != EACCES)) {
+    return -1;
+  }
+
+  // As a fallback, create a raw socket, which requires root permissions
+  // or CAP_NET_RAW to be granted to the VPN client.
+  m_socket = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP);
```

Comment:
I like this approach. And we can also have another fallback solution (the 3rd one) which is similar to what mullvad does: let's run the `ping` tool. See: https://github.com/mullvad/mullvadvpn-app/blob/master/talpid-core/src/ping_monitor/unix.rs#L82-L113